### PR TITLE
Add class method for preset memo wise

### LINKED ISSUE #86: Class method for `preset_memo_wise`

### PR DESCRIPTION
**Pull request by ms-ati**
Steps:
  * Refactor specs to extract shared examples
  * Add class-level specs for .preset_memo_wise
  * Implement class-level .preset_memp_wise

Closes #86

### CODE REVIEW

File: lib/memo_wise.rb [hunk 1]
```diff
@@ -351,6 +351,27 @@ def #{method_name}#{args_str}
...
+        fetch_key
+      ].each do |method_name|
+        # Like calling 'module_function', but original method stays public
+        target.define_singleton_method(
+          method_name,
+          MemoWise.instance_method(method_name)
+        )
+
+        # Make private the class method copies of private instance methods
+        unless MemoWise.public_method_defined?(method_name)
```
Comment (jemmaissroff) [inline lib/memo_wise.rb:370]: What's the case where `MemoWise.public_method_defined?(method_name)` would return `true`?

File: lib/memo_wise.rb [hunk 2]
```diff
@@ -559,16 +585,18 @@ def reset_memo_wise(method_name = nil, *args, **kwargs)
 
   # Validates that {.memo_wise} has already been called on `method_name`.
   def validate_memo_wised!(method_name)
+    klass = instance_of?(Class) ? singleton_class : self.class
```
Comment (jemmaissroff) [inline lib/memo_wise.rb:588]: Is it worth extracting this line into a `klass` method if we use it below too?

**Review by jemmaissroff — CHANGES_REQUESTED** [review]

File: lib/memo_wise.rb (hunk 1 above)
Comment (ms-ati) [reply lib/memo_wise.rb:370]: Hi Jemma! In the current code, some of these methods are private helpers and would return false. `preset_memo_wise` is the public method and returns true

File: lib/memo_wise.rb (hunk 2 above)
Comment (ms-ati) [reply lib/memo_wise.rb:588]: Planning to do extraction and refactoring when I repeat this pattern for `reset_memo_wise` if that’s ok?

**Review by jemmaissroff — APPROVED** [review]

File: lib/memo_wise.rb [hunk 3]
```diff
@@ -613,16 +639,19 @@ def reset_memo_wise(method_name = nil, *args, **kwargs)
...
       raise ArgumentError, "#{method_name} is not a memo_wised method"
     end
   end
 
   # Returns arguments key to lookup memoized results for given `method_name`.
   def fetch_key(method_name, *args, **kwargs)
-    method = self.class.instance_method(method_name)
+    klass = instance_of?(Class) ? singleton_class : self.class
+    method = klass.instance_method(method_name)
+    has_arg = MemoWise.has_arg?(method)
```
Comment (ms-ati) [inline lib/memo_wise.rb:654]: Oops, this looks like an accident during the rebase, will remove

**Pull request merged**